drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,2042 @@
+#define NUMAKER_USBD_GET_ADDR(USBD)              ((uint32_t)(USBD->FADDR))
+#define NUMAKER_USBD_ENABLE_INT(USBD, intr)      (USBD->INTEN |= (intr))
+#define NUMAKER_USBD_GET_INT_FLAG(USBD)          ((uint32_t)(USBD->INTSTS))
+#define NUMAKER_USBD_CLR_INT_FLAG(USBD, flag)    (USBD->INTSTS = (flag))
+#define NUMAKER_USBD_GET_EP_INT_FLAG(USBD)       ((uint32_t)(USBD->EPINTSTS))
+#define NUMAKER_USBD_CLR_EP_INT_FLAG(USBD, flag) (USBD->EPINTSTS = (flag))
+#define NUMAKER_USBD_GET_EP_FLAG(USBD)           ((uint32_t)(USBD->EPSTS))
+#define NUMAKER_USBD_GET_BUS_STATE(USBD)         ((uint32_t)(USBD->ATTR & 0x300f))
+#define NUMAKER_USBD_IS_ATTACHED(USBD)           ((USBD->VBUSDET & USBD_VBUSDET_VBUSDET_Msk))
+#define NUMAKER_USBD_STOP_TRANSACTION(USBD, ep)                                                    \
+	(*((__IO uint32_t *)((uint32_t)(&USBD->EP[0].CFGP) + (uint32_t)((ep) << 4))) |=            \
+	 USBD_CFGP_CLRRDY_Msk)
+#define NUMAKER_USBD_SET_DATA1(USBD, ep)                                                           \
+	(*((__IO uint32_t *)((uint32_t)(&USBD->EP[0].CFG) + (uint32_t)((ep) << 4))) |=             \
+	 USBD_CFG_DSQSYNC_Msk)
+#define NUMAKER_USBD_SET_DATA0(USBD, ep)                                                           \
+	(*((__IO uint32_t *)((uint32_t)(&USBD->EP[0].CFG) + (uint32_t)((ep) << 4))) &=             \
+	 (~USBD_CFG_DSQSYNC_Msk))
+#define NUMAKER_USBD_SET_PAYLOAD_LEN(USBD, ep, size)                                               \
+	(*((__IO uint32_t *)((uint32_t)(&USBD->EP[0].MXPLD) + (uint32_t)((ep) << 4))) = (size))
+#define NUMAKER_USBD_GET_PAYLOAD_LEN(USBD, ep)                                                     \
+	(*((__IO uint32_t *)((uint32_t)(&USBD->EP[0].MXPLD) + (uint32_t)((ep) << 4))))
+#define NUMAKER_USBD_CONFIG_EP(USBD, ep, config)                                                   \
+	(*((__IO uint32_t *)((uint32_t)(&USBD->EP[0].CFG) + (uint32_t)((ep) << 4))) = (config))
+#define NUMAKER_USBD_SET_EP_BUF_ADDR(USBD, ep, offset)                                             \
+	(*((__IO uint32_t *)((uint32_t)(&USBD->EP[0].BUFSEG) + (uint32_t)((ep) << 4))) = (offset))
+#define NUMAKER_USBD_GET_EP_BUF_ADDR(USBD, ep)                                                     \
+	(*((__IO uint32_t *)((uint32_t)(&USBD->EP[0].BUFSEG) + (uint32_t)((ep) << 4))))
+#define NUMAKER_USBD_SET_EP_STALL(USBD, ep)                                                        \
+	(*((__IO uint32_t *)((uint32_t)(&USBD->EP[0ul].CFGP) + (uint32_t)((ep) << 4))) |=          \
+	 USBD_CFGP_SSTALL_Msk)
+#define NUMAKER_USBD_CLR_EP_STALL(USBD, ep)                                                        \
+	(*((__IO uint32_t *)((uint32_t)(&USBD->EP[0].CFGP) + (uint32_t)((ep) << 4))) &=            \
+	 ~USBD_CFGP_SSTALL_Msk)
+#define NUMAKER_USBD_GET_EP_STALL(USBD, ep)                                                        \
+	(*((__IO uint32_t *)((uint32_t)(&USBD->EP[0].CFGP) + (uint32_t)((ep) << 4))) &             \
+	 USBD_CFGP_SSTALL_Msk)
```

Comment:
Please follow my suggestion in comment https://github.com/zephyrproject-rtos/zephyr/pull/66418#discussion_r1433854963, you can use inline function where appropriate.

Reply:
as suggested, expanded these macro functions